sympy/functions/special/beta_functions.py

```diff
@@ -118,3 +128,201 @@ def _eval_conjugate(self):
 
     def _eval_rewrite_as_gamma(self, x, y, piecewise=True, **kwargs):
         return self._eval_expand_func(**kwargs)
+
+    def _eval_rewrite_as_Integral(self, x, y, **kwargs):
+        from sympy.integrals.integrals import Integral
+        t = Dummy('t')
+        return Integral(t**(x - 1)*(1 - t)**(y - 1), (t, 0, 1))
+
+###############################################################################
+########################## INCOMPLETE BETA FUNCTION ###########################
+###############################################################################
+
+class betainc(Function):
+    r"""
+    The Generalized Incomplete Beta function is defined as
+
+    .. math::
+        \mathrm{B}_{(x_1, x_2)}(a, b) = \int_{x_1}^{x_2} t^{a - 1} (1 - t)^{b - 1} dt
+
+    The Incomplete Beta function is a special case
+    of the Generalized Incomplete Beta function :
+
+    .. math:: \mathrm{B}_z (a, b) = \mathrm{B}_{(0, z)}(a, b)
+
+    The Incomplete Beta function satisfies :
+
+    .. math:: \mathrm{B}_z (a, b) = (-1)^a \mathrm{B}_{\frac{z}{z - 1}} (a, 1 - a - b)
+
+    The Beta function is a special case of the Incomplete Beta function :
+
+    .. math:: \mathrm{B}(a, b) = \mathrm{B}_{1}(a, b)
+
+    Examples
+    ========
+
+    >>> from sympy import betainc, symbols, conjugate
+    >>> a, b, x, x1, x2 = symbols('a b x x1 x2')
+
+    The Generalized Incomplete Beta function is given by:
+
+    >>> betainc(a, b, x1, x2)
+    betainc(a, b, x1, x2)
+
+    The Incomplete Beta function can be obtained as follows:
+
+    >>> betainc(a, b, 0, x)
+    betainc(a, b, 0, x)
+
+    The Incomplete Beta function obeys the mirror symmetry:
+
+    >>> conjugate(betainc(a, b, x1, x2))
+    betainc(conjugate(a), conjugate(b), conjugate(x1), conjugate(x2))
+
+    We can numerically evaluate the Incomplete Beta function
+    to arbitrary precision on the whole complex plane:
+
+    >>> from sympy import betainc, I
+    >>> betainc(2, 3, 4, 5).evalf(10)
+    56.08333333
+    >>> betainc(0.75, 1 - 4*I, 0, 2 + 3*I).evalf(25)
+    0.2241657956955709603655887 + 0.3619619242700451992411724*I
+
+    The Generalized Incomplete Beta function can be expressed
+    in terms of the Generalized Hypergeometric function.
+
+    >>> from sympy import hyper
+    >>> betainc(a, b, x1, x2).rewrite(hyper)
+    (-x1**a*hyper((a, 1 - b), (a + 1,), x1) + x2**a*hyper((a, 1 - b), (a + 1,), x2))/a
+
+    See Also
+    ========
+
+    beta: Beta function
+    hyper: Generalized Hypergeometric function
+
+    References
+    ==========
+
+    .. [1] https://en.wikipedia.org/wiki/Beta_function#Incomplete_beta_function
+    .. [2] https://dlmf.nist.gov/8.17
+    .. [3] https://functions.wolfram.com/GammaBetaErf/Beta4/
+    .. [4] https://functions.wolfram.com/GammaBetaErf/BetaRegularized4/02/
+
+    """
+    nargs = 4
+    unbranched = True
+
+    def _eval_is_real(self):
+        return all(arg.is_real for arg in self.args)
```

Comment:
This will give False if the arguments are not known to be real. The `is_real` property is a fuzzy bool that can be `True`, `False`, or `None`. Functions for operating on ordinary bools such as `all` should not be used with fuzzy bool because they do not correctly handle None:
```python
In [13]: x = Symbol('x')

In [14]: betainc(2, 3, 4, x).is_real
Out[14]: False

In [15]: betainc(2, 3, 4, x).subs(x, 5).is_real
Out[15]: True
```
That should be fixed and a test added e.g.:
```
if all(arg.is_real for arg in self.args):
    return True
```
That will give None if any arg gives None or False. Returning None is fine though as it just means "unknown".

Is it necessarily the case that if any of the args is not real the function is not real? If so you could change `all` to fuzzy_and` here which would give `False` is any arg gives `False`.

Reply:
> Is it necessarily the case that if any of the args is not real the function is not real? If so you could change all to fuzzy_andhere which would giveFalseis any arg givesFalse`.

When `x1 = x2  = 0`, the integral would always be 0, so this is not necessarily true. I'll just change it to the if condition you suggested above.

Reply:
``` python
if all(arg.is_real for arg in self.args):
    return True
```
This will return None when I call `betainc(1, 2, -3*I, 4).is_real`. Is this intended?

Reply:
Assumptions queries use three-valued logic (TVL) and return True, False or None with None meaning "unknown". It is always logically consistent in TVL to return None even if the result of the query could correctly be True or False. On other hand returning True when the answer should be False or vice-versa is a bug that needs to be fixed. What this means is that we should be cautious and allow None to be returned by default  unless we are absolutely sure that the result is True (or False) given the result of any checks.

So yes when I wrote that I did intend for it to return None in that case. If we want that to return False and it is easy to introduce checks that would return False for this case without *ever* being incorrect then we can add further checking to make this return False. An assumptions query like this should be "cheap" though so only simple checking should be used.

Reply:
Sure, thanks for the clarification.